### PR TITLE
AF-2347: [wires-grids] CellContextUtilities.makeCellRenderContext() wrong for scrolled grids

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/AF-2347

### CODE REVIEW

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidget.java [hunk 1]
```diff
@@ -244,6 +248,12 @@ public double getHeight() {
         return height;
     }
 
+    private double getHeight(final BaseGridRendererHelper.RenderingInformation renderingInformation) {
+        double height = renderer.getHeaderHeight();
+        height = height + renderingInformation.getAllRowHeights().stream().reduce(0d, Double::sum);
```
Comment (gitgabrio) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidget.java]: @manstis 
pls replace with for-loop (or parallel stream) for performance

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidget.java [hunk 2]
```diff
@@ -291,17 +303,45 @@ protected void drawWithoutTransforms(Context2D context,
 
         if (!isSelectionLayer) {
             //If there's no RenderingInformation the GridWidget is not visible
+            if (LOGGER.isLoggable(Level.FINEST)) {
```
Comment (gitgabrio) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidget.java]: @manstis 
I understand reason of all those "Logger" (I had to do the same elsewhere) - but could you please just create a single method, invoked with the first part of the message to be logged (everything else is the same) ?

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/renderers/columns/impl/ColumnRenderingStrategyMerged.java [hunk 3]
```diff
@@ -305,10 +306,10 @@ protected static boolean isCollapsedCellMixedValue(final GridData model,
     }
 
     protected static double getCellHeight(final int rowIndex,
-                                          final GridData model,
+                                          final List<Double> allRowHeights,
                                           final GridCell<?> cell) {
         return IntStream.range(rowIndex, rowIndex + cell.getMergedCellCount())
```
Comment (gitgabrio) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/renderers/columns/impl/ColumnRenderingStrategyMerged.java]: @manstis 
could you please replace with for-loop (or parallel stream) for performance ?

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/renderers/grids/impl/BaseGridRendererHelper.java [hunk 4]
```diff
@@ -480,6 +536,7 @@ public RenderingInformation(final Bounds bounds,
             this.floatingBlockInformation = floatingBlockInformation;
             this.minVisibleRowIndex = minVisibleRowIndex;
             this.maxVisibleRowIndex = maxVisibleRowIndex;
+            this.allRowHeights = allRowHeights;
             this.visibleRowOffsets = visibleRowOffsets;
```
Comment (gitgabrio) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/renderers/grids/impl/BaseGridRendererHelper.java]: @manstis 
What's the difference between `allRowHeights` and `visibleRowOffsets` ? If the latter is just a subset of the former, can't we just save the limits of the offset without iterating twice to create two lists?

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidget.java (hunk 1 above)
Comment (manstis) [reply uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidget.java]: Done.

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/renderers/columns/impl/ColumnRenderingStrategyMerged.java (hunk 3 above)
Comment (manstis) [reply uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/renderers/columns/impl/ColumnRenderingStrategyMerged.java]: Done.

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/renderers/grids/impl/BaseGridRendererHelper.java (hunk 4 above)
Comment (manstis) [reply uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/renderers/grids/impl/BaseGridRendererHelper.java]: They're not quite the same.. `allRowHeights` contains the rows' heights e.g. (as an array to make it easier to understand what I'm saying) `{20, 20, 20}` for three rows of equal height. `visibleRowOffsets` contains subsequent row's Y-offset from 0 e.g. `{0, 20, 40}` for the same three rows as shown in `allRowHeights`. The latter only contains visible rows as it's used mainly for rendering the merged grid view.

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidget.java (hunk 2 above)
Comment (manstis) [reply uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/main/java/org/uberfire/ext/wires/core/grids/client/widget/grid/impl/BaseGridWidget.java]: Done.

**Comment by manstis** [comment]
@gitgabrio I've made the changes you requested (on PR #843)...

It is only the last commit on this PR that needs reviewing for this PR.. but I'll ping you again when the grid performance PRs have been merged and this one reviewed......

**Comment by manstis** [comment]
@gitgabrio @jomarko This PR is ready for review...

@gitgabrio Further to your original comment this PR now only contains the fix for showing the context menu for cells in the correct place follow a scroll of the grid.

**Comment by manstis** [comment]
@danielezonca New test added for the _scrolled_ scenario.

**Comment by jomarko** [comment]
jenkins execute full downstream build

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/test/java/org/uberfire/ext/wires/core/grids/client/util/CellContextUtilitiesTest.java [hunk 5]
```diff
@@ -84,9 +84,16 @@
...
+    private BaseGridRow row1 = new BaseGridRow();
+    private BaseGridRow row2 = new BaseGridRow();
+    private BaseGridRow row3 = new BaseGridRow();
 
     @Before
     public void setup() {
+        row1 = new BaseGridRow();
+        row2 = new BaseGridRow();
+        row3 = new BaseGridRow();
+
```
Comment (jomarko) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/test/java/org/uberfire/ext/wires/core/grids/client/util/CellContextUtilitiesTest.java]: Please initialize either in before or in field initialization.

File: uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/test/java/org/uberfire/ext/wires/core/grids/client/util/CellContextUtilitiesTest.java [hunk 6]
```diff
@@ -170,15 +174,52 @@ public void testMakeCellRenderContextOneRow() {
...
-                .isEqualTo(headerRowsHeight);
+                .isEqualTo(HEADER_HEIGHT);
     }
 
     @Test
     public void testMakeCellRenderContextThreeRows() {
-        final double headerRowsHeight = 100.0;
-        final BaseGridRow row1 = new BaseGridRow();
-        final BaseGridRow row2 = new BaseGridRow();
-        final BaseGridRow row3 = new BaseGridRow();
```
Comment (jomarko) [inline uberfire-extensions/uberfire-wires/uberfire-wires-core/uberfire-wires-core-grids/src/test/java/org/uberfire/ext/wires/core/grids/client/util/CellContextUtilitiesTest.java]: few more tests create `new BaseGridRow()` as part of test method, do you want to update them to use `row1`, `row2` ...?

**Comment by manstis** [comment]
@jomarko Updates made.

**Comment by manstis** [comment]
@danielezonca @jomarko @gitgabrio Could you please approve this PR when you get a chance.

**Pull request merged**